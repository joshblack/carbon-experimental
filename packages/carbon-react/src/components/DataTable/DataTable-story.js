import React, { Component } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import DataTableRow from '../DataTableRow';
import DataTableData from '../DataTableData';
import {
  DataTable,
  DataTableToolbar,
  DataTableToolbarContent,
  DataTableToolbarAction,
  DataTableHead,
  DataTableColumnHeader,
  DataTableBody,
  DataTableBatchActions,
  DataTableActionList,
  DataTableBatchAction,
  DataTableSearch,
  DataTableContainer,
  DataTableSelectAll,
} from '../DataTable';
import PaginationV2 from '../PaginationV2';
import Button from '../Button';
import OverflowMenu from '../OverflowMenu';
import OverflowMenuItem from '../OverflowMenuItem';

const paginationProps = {
  pageSizes: [10, 20, 30, 40, 50],
};

const initialRows = [
  {
    name: 'Load Balancer 3',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Kevins VM Groups',
    status: 'Active',
  },
  {
    name: 'Load Balancer 1',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Maureens VM Groups',
    status: 'Active',
  },
  {
    name: 'Load Balancer 2',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Andrews VM Groups',
    status: 'Active',
  },
];

const initialExpandedRows = [
  {
    rowContent: {
      name: 'Load Balancer 3',
      protocol: 'HTTP',
      something: '80',
      rule: 'Round Robin',
      attached_groups: 'Kevins VM Groups',
      status: 'Active',
    },
    expandedRowContent: {
      html: `<div>
      <h1>Expandable Row Content 1</h1>
      <p>Description here.</p>
    </div>`,
    },
  },
  {
    rowContent: {
      name: 'Load Balancer 1',
      protocol: 'HTTP',
      something: '80',
      rule: 'Round Robin',
      attached_groups: 'Maureens VM Groups',
      status: 'Active',
    },
    expandedRowContent: {
      html: `<div>
      <h1>Expandable Row Content 2</h1>
      <p>Description here.</p>
    </div>`,
    },
  },
  {
    rowContent: {
      name: 'Load Balancer 2',
      protocol: 'HTTP',
      something: '80',
      rule: 'Round Robin',
      attached_groups: 'Andrews VM Groups',
      status: 'Active',
    },
    expandedRowContent: {
      html: `<div>
      <h1>Expandable Row Content 3</h1>
      <p>Description here.</p>
    </div>`,
    },
  },
];

const headers = Object.keys(initialRows[0]).map(key => ({
  key,
  title: key.charAt(0).toUpperCase() + key.substring(1),
}));

class BasicDataTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      checked: [],
      selectAll: false,
      rows: this.props.rows,
      filteredRows: this.props.rows,
    };
  }

  static propTypes = {
    zebra: PropTypes.bool,
    rows: PropTypes.array,
    isSelectable: PropTypes.bool,
  };

  static defaultProps = {
    zebra: true,
    rows: initialRows,
    isSelectable: true,
  };

  selectAll = () => {
    const checked = this.state.checked;
    this.state.rows.forEach((row, index) => {
      if (this.state.selectAll) {
        checked[index] = false;
      } else {
        checked[index] = true;
      }
    });
    this.setState({
      checked,
      selectAll: !this.state.selectAll,
    });
  };

  selectRow = index => {
    const checked = this.state.checked;
    checked[index] = checked[index] ? !checked[index] : true;
    this.setState({
      checked,
      selectAll: checked[index] === false ? false : this.state.selectAll,
    });
  };

  clearAll = () => {
    const checked = this.state.checked;
    this.state.rows.forEach((row, index) => {
      checked[index] = checked[index] ? !checked[index] : false;
    });
    this.setState({
      checked,
      selectAll: false,
    });
  };

  getCheckedItems = () => {
    const checked = this.state.checked;
    let checkedItems = 0;
    checked.forEach(entry => {
      if (entry) {
        checkedItems++;
      }
    });
    return checkedItems;
  };

  searchTable = evt => {
    let newRows = [];
    this.state.rows.map(obj => {
      Object.keys(obj).forEach(key => {
        if (
          obj[key]
            .toUpperCase()
            .includes(evt.currentTarget.value.toUpperCase()) &&
          !newRows.includes(obj)
        ) {
          newRows.push(obj);
        }
      });
    });
    this.setState({
      filteredRows: newRows,
    });
  };

  render() {
    const checkedItems = this.getCheckedItems();
    const showBatchActions = checkedItems > 0;
    const tableClasses = classNames({
      'bx--data-table-v2': true,
      'bx--data-table-v2--zebra': this.props.zebra,
    });
    const { isSelectable } = this.props;
    return (
      <div>
        <DataTableContainer title="Table title">
          <DataTableToolbar>
            <DataTableBatchActions
              totalSelected={checkedItems}
              showBatchActions={showBatchActions}
              handleClick={this.clearAll}>
              <DataTableActionList>
                <DataTableBatchAction onClick={action('Batch Action 1')}>
                  Ghost
                </DataTableBatchAction>
                <DataTableBatchAction onClick={action('Batch Action 2')}>
                  Ghost
                </DataTableBatchAction>
                <DataTableBatchAction onClick={action('Batch Action 3')}>
                  Ghost
                </DataTableBatchAction>
              </DataTableActionList>
            </DataTableBatchActions>
            <DataTableSearch
              onInput={this.searchTable}
              onChange={this.searchTable}
            />
            <DataTableToolbarContent>
              <DataTableToolbarAction
                iconName="download"
                iconDescription="Download"
                onClick={action('Toolbar Action 1')}
              />
              <DataTableToolbarAction
                iconName="edit"
                iconDescription="Edit"
                onClick={action('Toolbar Action 1')}
              />
              <DataTableToolbarAction
                iconName="settings"
                iconDescription="Settings"
                onClick={action('Toolbar Action 1')}
              />
              <Button onClick={action('Add new row')} kind="primary">
                Add new
              </Button>
            </DataTableToolbarContent>
          </DataTableToolbar>
          <DataTable
            initialRows={initialRows}
            headers={headers}
            render={({ headers, getHeaderProps }) => (
              <table className={tableClasses}>
                <DataTableHead>
                  <DataTableRow>
                    {isSelectable && (
                      <DataTableSelectAll
                        checked={this.state.selectAll}
                        onClick={this.selectAll}
                      />
                    )}
                    {headers.map(header => (
                      <DataTableColumnHeader
                        key={header.key}
                        sortable={true}
                        {...getHeaderProps(header)}>
                        {header.title}
                      </DataTableColumnHeader>
                    ))}
                    <DataTableColumnHeader />
                  </DataTableRow>
                </DataTableHead>
                <DataTableBody>
                  {this.state.filteredRows.map((row, i) => {
                    return (
                      <DataTableRow key={`row${i}`}>
                        <DataTableData
                          onClick={() => this.selectRow(i)}
                          key={`a${i}`}
                          isSelectable={true}
                          isSelected={
                            this.state.checked[i]
                              ? this.state.checked[i]
                              : false
                          }
                        />
                        {Object.keys(row).map((rowData, j) => {
                          return (
                            <DataTableData key={`rowdata${j}`}>
                              {row[rowData]}
                            </DataTableData>
                          );
                        })}
                        <DataTableData
                          className="bx--table-overflow"
                          key={`c${i}`}>
                          <OverflowMenu flipped floatingMenu>
                            <OverflowMenuItem itemText="Option 1" />
                            <OverflowMenuItem itemText="Option 2" />
                            <OverflowMenuItem itemText="Option 3" />
                            <OverflowMenuItem itemText="Option 4" />
                            <OverflowMenuItem
                              itemText="Danger option"
                              hasDivider
                              isDelete
                            />
                          </OverflowMenu>
                        </DataTableData>
                      </DataTableRow>
                    );
                  })}
                </DataTableBody>
              </table>
            )}
          />
        </DataTableContainer>
        <PaginationV2 {...paginationProps} totalItems={50} />
      </div>
    );
  }
}

class ExpandableDataTable extends Component {
  state = {
    expanded: [],
    selectAll: false,
    rows: this.props.rows,
  };

  static propTypes = {
    zebra: PropTypes.bool,
    rows: PropTypes.array,
  };

  static defaultProps = {
    zebra: true,
    rows: initialExpandedRows,
  };

  createNewRow = (content, parent) => {
    const newRow = document.createElement('tr');
    const colspan = Object.keys(this.state.rows[0].rowContent).length + 2;
    parent.dataset.parentRow = '';
    newRow.classList.add('bx--expandable-row-v2');
    newRow.dataset.childRow = '';
    newRow.innerHTML = `
      <td colspan=${colspan}>
        ${content}
      </td>
    `;
    Object.keys(parent.parentElement.children).map(child => {
      if (parent.parentElement.children[child] === parent) {
        parent.parentElement.insertBefore(
          newRow,
          parent.parentElement.childNodes[+child + 1]
        );
        newRow.addEventListener('mouseover', () => {
          parent.classList.add('bx--expandable-row--hover-v2');
        });
        newRow.addEventListener('mouseout', () => {
          parent.classList.remove('bx--expandable-row--hover-v2');
        });
      }
    });
  };

  removeRow = parent => {
    Object.keys(parent.parentElement.children).map(child => {
      if (parent.parentElement.children[child] === parent) {
        parent.parentElement.removeChild(
          parent.parentElement.children[+child + 1]
        );
      }
    });
  };

  expandRow = (index, evt) => {
    const parent = evt.currentTarget.parentElement;
    const expanded = this.state.expanded;
    const rows = this.state.rows;
    if (expanded[index] === true) {
      expanded[index] = false;
    } else {
      expanded[index] = expanded[index] ? !expanded[index] : true;
    }
    this.setState({
      expanded,
    });
    if (this.state.expanded[index]) {
      this.createNewRow(rows[index].expandedRowContent.html, parent);
    } else {
      this.removeRow(parent);
    }
  };

  rowHover = evt => {
    evt.currentTarget.classList.add('bx--expandable-row--hover-v2');
  };

  removeRowHover = evt => {
    evt.currentTarget.classList.remove('bx--expandable-row--hover-v2');
  };

  render() {
    const tableClasses = classNames({
      'bx--data-table-v2': true,
      'bx--data-table-v2--zebra': false,
    });

    return (
      <DataTableContainer title="Table title">
        <DataTable
          initialRows={initialExpandedRows}
          headers={headers}
          render={({ rows, headers }) => (
            <table className={tableClasses}>
              <DataTableHead>
                <DataTableRow>
                  <th />
                  {headers.map(header => (
                    <th key={header.key}>{header.title}</th>
                  ))}
                  <th />
                </DataTableRow>
              </DataTableHead>
              <DataTableBody>
                {rows.map((row, i) => {
                  return (
                    <DataTableRow
                      className={classNames({
                        'bx--parent-row-v2': true,
                        'bx--expandable-row-v2': this.state.expanded[i],
                      })}>
                      <DataTableData
                        onClick={evt => {
                          this.expandRow(i, evt);
                        }}
                        key={`a${i}`}
                        isExpandable={true}
                        isExpanded={
                          this.state.expanded[i]
                            ? this.state.expanded[i]
                            : false
                        }
                      />
                      {Object.keys(row.rowContent).map((content, j) => {
                        return (
                          <DataTableData key={`rowdata${j}`}>
                            {row.rowContent[content]}
                          </DataTableData>
                        );
                      })}
                      <DataTableData
                        className="bx--table-overflow"
                        key={`c${i}`}>
                        <OverflowMenu flipped floatingMenu>
                          <OverflowMenuItem itemText="Option 1" />
                          <OverflowMenuItem itemText="Option 2" />
                          <OverflowMenuItem itemText="Option 3" />
                          <OverflowMenuItem itemText="Option 4" />
                          <OverflowMenuItem
                            itemText="Danger option"
                            hasDivider
                            isDelete
                          />
                        </OverflowMenu>
                      </DataTableData>
                    </DataTableRow>
                  );
                })}
              </DataTableBody>
            </table>
          )}
        />
      </DataTableContainer>
    );
  }
}

storiesOf('DataTable', module)
  .addWithInfo(
    'Data Table',
    `
      Data table
    `,
    () => <BasicDataTable />
  )
  .addWithInfo(
    'Expandable table',
    `
      Expandable table
    `,
    () => <ExpandableDataTable />
  );
