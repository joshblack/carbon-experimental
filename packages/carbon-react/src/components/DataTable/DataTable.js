import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import Button from '../Button';
import Icon from '../Icon';
import Search from '../Search';

const toggleSortDirection = direction => {
  if (direction === 'DESC') {
    return 'ASC';
  }

  return 'DESC';
};

const sortRow = (key, direction) => (a, b) => {
  if (direction === 'DESC') {
    return a[key].localeCompare(b[key]);
  }

  return b[key].localeCompare(a[key]);
};

export class DataTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      rows: props.initialRows,
      sortKey: null,
      sortDirection: null,
    };
    this.sortBy = this.sortBy.bind(this);
    this.getHeaderProps = this.getHeaderProps.bind(this);
  }

  sortBy(key) {
    return () => {
      const direction =
        this.state.sortKey === key
          ? toggleSortDirection(this.state.sortDirection)
          : 'DESC';

      this.setState(state => {
        return {
          rows: state.rows.sort(sortRow(key, direction)),
          sortKey: key,
          sortDirection: direction,
        };
      });
    };
  }

  getHeaderProps(header) {
    return {
      onClick: this.sortBy(header.key),
      onKeyDown: () => {},
    };
  }

  render() {
    const { render, children, headers } = this.props;
    const { rows } = this.state;

    const renderProps = {
      rows,
      headers,
      sortBy: this.sortBy,
      getHeaderProps: this.getHeaderProps,
    };

    if (render !== undefined) {
      return render(renderProps);
    }

    if (children !== undefined) {
      return children(renderProps);
    }

    return null;
  }
}

export class DataTableColumnHeader extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sorted: false,
      active: false,
    };
  }

  setActive = () => {
    this.setState({
      active: true,
    });
  };

  removeActive = () => {
    this.setState({
      active: false,
    });
  };

  handleClick = () => {
    this.setState(state => ({
      sorted: !state.sorted,
    }));
  };

  render() {
    const { children, className, sortable, ...other } = this.props;
    const tableSortClasses = classNames('bx--table-sort-v2', {
      'bx--table-sort-v2--active': this.state.active,
      'bx--table-sort-v2--ascending': this.state.sorted,
    });
    return (
      <th {...other} className={className}>
        {sortable ? (
          <button
            onFocus={this.setActive}
            onClick={this.handleClick}
            className={tableSortClasses}>
            {children}
            <Icon
              className="bx--table-sort-v2__icon"
              name="caret--down"
              description="Sort arrow"
            />
          </button>
        ) : (
          children
        )}
      </th>
    );
  }
}

DataTableColumnHeader.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  sortable: PropTypes.bool,
  onClick: PropTypes.func,
};

export const DataTableHead = ({ ...props }) => {
  const { children, className, ...other } = props;

  return (
    <thead {...other} className={className}>
      {children}
    </thead>
  );
};

DataTableHead.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export const DataTableBody = ({ ...props }) => {
  const { className, children, ...other } = props;

  return (
    <tbody {...other} className={className}>
      {children}
    </tbody>
  );
};

DataTableBody.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export const DataTableToolbarContent = ({ ...props }) => {
  const { className, children, ...other } = props;

  const toolbarContentClasses = classNames(className, 'bx--toolbar-content');

  return (
    <div className={toolbarContentClasses} {...other}>
      {children}
    </div>
  );
};

DataTableToolbarContent.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export const DataTableToolbarAction = ({ ...props }) => {
  const { className, iconName, iconDescription, ...other } = props;

  const toolbarActionClasses = classNames(className, 'bx--toolbar-action');

  return (
    <button className={toolbarActionClasses} {...other}>
      <Icon
        className="bx--toolbar-action__icon"
        name={iconName}
        description={iconDescription}
      />
    </button>
  );
};

DataTableToolbarAction.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  iconName: PropTypes.string,
  iconDescription: PropTypes.string,
};

export const DataTableToolbar = ({ ...props }) => {
  const { className, children, ...other } = props;

  const toolbarClasses = classNames(className, 'bx--table-toolbar');

  return (
    <section className={toolbarClasses} {...other}>
      {children}
    </section>
  );
};

DataTableToolbar.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export const DataTableActionList = ({ ...props }) => {
  const { className, children, ...other } = props;

  const actionListClasses = classNames(className, 'bx--action-list');

  return (
    <div className={actionListClasses} {...other}>
      {children}
    </div>
  );
};

DataTableActionList.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export const DataTableBatchAction = ({ ...props }) => {
  const { className, children, ...other } = props;

  return (
    <Button
      small
      kind="ghost"
      className={className}
      icon="add--glyph"
      iconDescription="Add"
      {...other}>
      {children}
    </Button>
  );
};

DataTableBatchAction.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export const DataTableBatchActions = ({ ...props }) => {
  const {
    className,
    children,
    showBatchActions,
    totalSelected,
    handleClick,
    ...other
  } = props;

  const batchActionsClasses = classNames(
    {
      'bx--batch-actions': true,
      'bx--batch-actions--active': showBatchActions,
    },
    className
  );

  return (
    <div className={batchActionsClasses} {...other}>
      {children}
      <div className="bx--batch-summary">
        <p className="bx--batch-summary__para">
          <span>{totalSelected}</span>{' '}
          {totalSelected > 1 ? 'items selected' : 'item selected'}
        </p>
        <button className="bx--batch-summary__cancel" onClick={handleClick}>
          Cancel
        </button>
      </div>
    </div>
  );
};

DataTableBatchAction.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  showBatchActions: PropTypes.bool,
  totalSelected: PropTypes.number,
  handleClick: PropTypes.func,
};

export class DataTableContainer extends Component {
  render() {
    const tableContainerClasses = classNames(
      className,
      'bx--data-table-v2-container'
    );
    const { children, className, title, ...other } = this.props;
    return (
      <div className={tableContainerClasses} {...other}>
        <h4 className="bx--data-table-v2-header">{title}</h4>
        {children}
      </div>
    );
  }
}

DataTableContainer.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  containerClassName: PropTypes.string,
  title: PropTypes.string,
};

export const DataTableSearch = props => {
  const { className, searchContainerClass, onChange, ...other } = props;

  const searchContainerClasses = classNames(
    searchContainerClass,
    'bx--toolbar-search-container'
  );

  return (
    <div className={searchContainerClasses}>
      <Search
        className={className}
        {...other}
        small
        id="search-2"
        labelText="Filter table"
        placeHolderText="Search"
        onChange={onChange}
      />
    </div>
  );
};

DataTableSearch.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  searchContainerClasses: PropTypes.string,
};

export const DataTableSelectAll = props => {
  const { className, onClick, checked, ...other } = props;

  const handleClick = () => {
    onClick();
  };

  return (
    <th>
      <input
        checked={checked}
        onClick={handleClick}
        type="checkbox"
        className={classNames(className, 'bx--checkbox')}
        {...other}
        value="green"
        name="checkbox-20"
        id="bx--checkbox-20"
      />
      <label htmlFor="bx--checkbox-20" className="bx--checkbox-label">
        <span className="bx--checkbox-appearance">
          <svg
            className="bx--checkbox-checkmark"
            width="12"
            height="9"
            viewBox="0 0 12 9"
            fillRule="evenodd">
            <path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z" />
          </svg>
        </span>
      </label>
    </th>
  );
};

DataTableSelectAll.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  onClick: PropTypes.func,
  checked: PropTypes.bool,
};
