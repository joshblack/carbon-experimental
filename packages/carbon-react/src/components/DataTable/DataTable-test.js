import React from 'react';
import { DataTableContainer } from '../DataTable';
import { shallow } from 'enzyme';

describe('DataTable', () => {
  it('should render', () => {
    const wrapper = shallow(
      <DataTableContainer>
        <div>Children</div>
      </DataTableContainer>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('has the expected classes', () => {
    const wrapper = shallow(
      <DataTableContainer>
        <div>Children</div>
      </DataTableContainer>
    );
    expect(wrapper.hasClass('bx--data-table-v2-container')).toEqual(true);
  });
});
