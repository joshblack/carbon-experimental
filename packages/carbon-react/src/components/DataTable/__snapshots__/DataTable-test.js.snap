// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render 1`] = `
<div
  className="bx--data-table-v2-container"
>
  <h4
    className="bx--data-table-v2-header"
  />
  <div>
    Children
  </div>
</div>
`;
